app/services/credits/applied_coupon_service.rb

```diff
@@ -57,5 +56,31 @@ def remaining_amount
       already_applied_amount = applied_coupon.credits.sum(:amount_cents)
       @remaining_amount = applied_coupon.amount_cents - already_applied_amount
     end
+
+    def should_terminate_applied_coupon?(credit_amount)
+      if applied_coupon.once?
+        applied_coupon.coupon.percentage? || credit_amount >= remaining_amount
+      else
+        applied_coupon.frequency_duration <= 0
+      end
+    end
+
+    def save_frequency_duration
+      return if applied_coupon.once?
```

Comment:
it's strange to have this test here, we should call the save_frequency_duration only for a recurring coupon, wdyt?